### ISSUE TITLE
CS8655 (exhaustive enum) is not triggered for nullable num value

### ISSUE THREAD

**Issue by juwens** [dotnet/roslyn#76817]
**Version Used**: C#12

**Steps to Reproduce**:

1. compile the sample code or test it in [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzguATEBqAPgYgAcAnCAc0gAIB3CYmWMyjKFCYAGwFNKBhAZQAcAVgBMAFgCwAKBlcYAVzCUAsgE8AoouUBvSgEFKAXxkyAAqL4ydMyncpnxqgBQBKW/ZvT7PysAD2/hwA/H6UALyUihwcANwevpQJvgD6EWEo1FAIAMYAFpQA9IWUADyROf7ExFw5CBxqNHQwKJTs/goIlGCsKIxRCjE02QUCggBswsLJPl6JiQjECjzhAHyUAIxwM/MAZhAcKCvrotve80bx54k7dreqmtqhAB7p0XH392mRr5nZ+UUSuVmP4uK0AHIAeQAKk16G0Ap1ur1+u9hghRkJJtNrr45vMfOotEoAHSGNaUABE+kpZwJl2SJmkRiAA)

```
#pragma warning disable CS8524

enum MyEnum { A }

class C
{
    void M()
    {
        bool? b = null;
        
        _ = b switch // <= correctly warns about missing null with CS8655
        {
            true => 1,
            false => 2,
        };
        
        
        MyEnum? x = null;
        
        _ = x switch // <= does NOT warn about missing null with CS8655
        {
            MyEnum.A => "A",
        };
    }
}
```

**Diagnostic Id**: CS8655: The switch expression does not handle some null inputs (it is not exhaustive).

If this is a report about a bug in an analyzer, please include the diagnostic ID and message if possible (e.g. `"IDE0030: Use coalesce expression"`).

**Expected Behavior**: nullable enum-variable switch-expression with missing null should yield CS8655

**Actual Behavior**: nullable enum-variable switch-expression with missing null does NOT yield CS8655

**Comment by jcouv** [comment]
The language allows an explicit conversion from an `int` to an enum type, so nothing guarantees that the enum value is limited to `MyEnum.A`.
```
MyEnum x = (MyEnum)42;
System.Console.Write(x); // prints 42

enum MyEnum { A = 0 }
```
[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4FgBQBZATwFEZUwACAD3IF5yAKQksgSgBYAmAblwAEBGAHQBhAPYwAzqIA2AU0EB1ZFASz6lFjxy5ZpCkz3kA3uQCCtcgAZyAX1xA)

**Comment by jjonescz** [comment]
@jcouv I don't think that's the issue here. The problem is that CS8655 is reported for `bool?` but not `MyEnum?`. CS8524 is reported for `MyEnum?` for the reason you describe. So perhaps the compiler does not report CS8655 anymore to avoid "repeating itself". But if customers suppress CS8524 (because they don't want to be warned about missing enum cases because don't cast ints to enums for example), they won't get the CS8655 warning there...

**Comment by juwens** [comment]
> [@jcouv](https://github.com/jcouv) I don't think that's the issue here. The problem is that CS8655 is reported for `bool?` but not `MyEnum?`. CS8524 is reported for `MyEnum?` for the reason you describe. So perhaps the compiler does not report CS8655 anymore to avoid "repeating itself". But if customers suppress CS8524 (because they don't want to be warned about missing enum cases because they know they don't cast ints to enums), they won't get the CS8655 warning there...

correct. CS8524 isn't involved with the null-check at all. With CS8524 enabled, one should see 8655 and 8524, cause CS8524 only indicates "malicious" `(MyEnum)42` values, but not null.
